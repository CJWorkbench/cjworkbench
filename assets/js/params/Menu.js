import React from 'react'
import PropTypes from 'prop-types'
import { MaybeLabel } from './util'

export default class MenuParam extends React.PureComponent {
  static propTypes = {
    name: PropTypes.string.isRequired,
    items: PropTypes.string, // like 'Apple|Banana|Kitten' -- DEPRECATED
    enumOptions: PropTypes.arrayOf(
      PropTypes.oneOfType([
        PropTypes.oneOf([ 'separator' ]),
        PropTypes.shape({
          value: PropTypes.any.isRequired,
          label: PropTypes.string.isRequired
        }).isRequired
      ]).isRequired
    ), // new-style menu -- once we nix "items" ("menu_items" in spec), add .isRequired here
    value: PropTypes.oneOfType([
      PropTypes.number.isRequired, // DEPRECATED
      PropTypes.string.isRequired // new-style menu
    ]).isRequired,
    isReadOnly: PropTypes.bool.isRequired,
    onChange: PropTypes.func.isRequired // onChange(newIndex) => undefined
  }

  onChange = (ev) => {
    const { items, onChange } = this.props

    // If "items" is set, this is a "deprecated_menu": values are integers.
    const value = items ? +ev.target.value : ev.target.value
    onChange(value)
  }

  get enumOptions () {
    const { items, enumOptions } = this.props

    if (items) {
      return items.split('|').map((label, value) => { // value is the index
        if (!label) return 'separator' // empty string is a separator
        return { label, value }
      })
    } else {
      return enumOptions
    }
  }

  render() {
    const { items, name, fieldId, label, isReadOnly, value } = this.props

    const options = this.enumOptions.map((option, i) => {
      if (option === 'separator') {
        return <option disabled key={i} className='separator' />
      } else {
        return <option key={option.value} value={option.value}>{option.label}</option>
      }
    })

    return (
      <React.Fragment>
        <MaybeLabel fieldId={fieldId} label={label} />
        <select
          name={name}
          id={fieldId}
          value={value}
          onChange={this.onChange}
          disabled={isReadOnly}
        >
          {options}
        </select>
      </React.Fragment>
    )
  }
}
