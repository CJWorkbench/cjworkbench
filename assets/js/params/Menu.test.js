import React from 'react'
import { shallow } from 'enzyme'
import Menu from './Menu';

describe('Menu', () => {
  it('matches snapshot', () => {
    const w = shallow(
      <Menu
        enumOptions={[{label: 'Apple', value: 'apple'}, {label: 'Kittens', value: 'kittens'}, {label: 'Banana', value: 'banana'}]}
        name='somename'
        value={'kittens'}
        onChange={jest.fn()}
        isReadOnly={false}
      />
    )
    expect(w).toMatchSnapshot()
  })
})
