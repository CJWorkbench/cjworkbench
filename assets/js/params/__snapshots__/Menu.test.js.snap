// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu matches snapshot 1`] = `
<Fragment>
  <MaybeLabel />
  <select
    disabled={false}
    name="somename"
    onChange={[Function]}
    value="1"
  >
    <option
      key="0"
      value="0"
    >
      Apple
    </option>
    <option
      key="1"
      value="1"
    >
      Kittens
    </option>
    <option
      key="2"
      value="2"
    >
      Banana
    </option>
  </select>
</Fragment>
`;
